packages/algorand-js/src/interpreter/parser.ts

```diff
@@ -0,0 +1,180 @@
+        default:
+          break;
+      }
+      i++;
+      continue;
+    }
+    if (!inString) {
+      const value = line.substr(start, i - start);
+      fields.push(value);
+      if (value === "base64" || value === "b64") {
+        inBase64 = true;
+      } else if (inBase64) {
+        inBase64 = false;
+      }
+    }
+    i++;
+
+    if (!inString) {
+      while (i < line.length && line[i] === ' ') {
+        i++;
+      }
+      start = i;
+    }
+  }
+
+  // add rest of the string if any
+  if (start < line.length) {
+    fields.push(line.substr(start, i - start));
+  }
+
+  return fields;
+}
+
+/**
+ * Description: Returns Opcode objects for fields with Pseudo-Ops
+ * @param fields : fields with field[0] as one of the Pseudo-Ops
+ */
+export function pseudoOps (fields: string[]): Operator {
+  switch (fields[0]) {
+    case "addr":
+      assertFieldLen(fields.length, 2);
+      if (!isValidAddress(fields[1])) {
+        throw new TealError(ERRORS.TEAL.INVALID_ADDR);
+      }
+
+      return new Addr(fields[1]);
+    case "int":
+      // allowed fields int 123 only. (int(12) is not allowed)
+      assertFieldLen(fields.length, 2);
+      assertOnlyDigits(fields[1]);
+
+      return new Int(BigInt(fields[1]));
+    default:
+      throw new TealError(ERRORS.TEAL.INVALID_OP_ARG);
+  }
+}
+
+/**
+ * Description: Returns Opcode objects for fields with Arithmetic-Opcodes
+ * @param fields : fields with field[0] as one of the Arithmetic-Opcodes
+ */
+export function arithmeticOps (fields: string[]): Operator {
+  assertFieldLen(fields.length, 1);
+
+  switch (fields[0]) {
+    case "+":
+      return new Add();
+    case "-":
+      return new Sub();
+    case "/":
+      return new Div();
+    case "*":
+      return new Mul();
+    default:
+      throw new TealError(ERRORS.TEAL.INVALID_OP_ARG);
+  }
+}
+
+/**
+ * Description: Returns Opcode object for given field
+ * @param fields : fields extracted from line
+ */
+export function opcodeFromFields (fields: string[]): Operator {
+  if (PSEUDO_OPS.includes(fields[0])) {
+    return pseudoOps(fields);
+  } else if (ARITHMETIC_OPERATIONS.includes(fields[0])) {
+    return arithmeticOps(fields);
+  } else {
+    throw new TealError(ERRORS.TEAL.INVALID_OP_ARG);
+  }
+}
+
+/**
+ * Description: Returns a list of Opcodes object after reading text from given TEAL file
+ * @param filename : Name of the TEAL file with location
+ */
+export async function parser (filename: string): Promise<Operator[]> {
+  const opCodeList = [] as Operator[];
+  const rl = readline.createInterface({
+    input: fs.createReadStream(filename), // file location
+    output: process.stdout,
+    terminal: false
+  });
+
+  for await (const line of rl) {
+    if (!(line.length === 0 || line.startsWith("//") || line.startsWith("#pragma"))) {
+      const fields = fieldsFromLine(line);
+
+      if (fields.length !== 0) {
+        opCodeList.push(opcodeFromFields(fields));
+      }
+    }
+  }
+  return opCodeList;
+}
```

Comment:
Let's add more structure to the parser.
This is how I would like to have it working (pseudo-code):

Firstly let's define a helper functions. In the code below `program` is a list of lines (strings)
```
function cleanup_line(line): {
  line = remove_comment(line)
  return trim_whitespaces(line)
}

function read_pragma(program): [program, pragma || undefined] {
  for line in program:
    line = cleanup_line(line)
    if line.startsWith("#pragma"):
      return program, new Pragma(...)
    else
      program.prepend(line) // prepend back the line, because it contains an instruction
      return program, undefined
}
```

Then we have to have a map of all op codes:

```
const opCodes = {
  add = Add
}
```
And finally the parser:

```
function parser(program: string[]) {
  [program, pragma] = read_pragma(program)
  opcodes = []
  for line in program:
      line = cleanup_line(line)
      line = trim_whitespaces(line)
      if line === "": continue
    
      words = line.split(" ")
      let op = opCodes[words[0]]
      if op === undefined:
         throw UnknownOpcode(words[0], )
      opcodes.push(new op(words.slice(1)...)
  return opcodes
}
```

maybe we should also add a line counter and use it in all error messages to make it clear in which line there is an error.

Reply:
It's just an idea. Let me know what do  you think. With this we can move more parsing logic to the operator constructor, and the parser implementation will be more clear.

Reply:
Yes, adding more structure will be good.
1. I think we should keep `fieldsFromLine` function because if we use `clean_up`. There is a case where in between spaces will be more than one, Ex: `byte        base64   Kkjh/=` then line.split(" ") will give in undesired result. There maybe more corner cases.

The structure will be as above with one difference:
```
function fieldsFromLine(line): {
       trim_whitespace(line)
       ...........
       ...........
       return fields.
}

function read_pragma(program): [program, pragma || undefined] {
  for line in program:
    line = cleanup_line(line)
    if line.startsWith("#pragma"):
      return program, new Pragma(...)
    else
      program.prepend(line) // prepend back the line, because it contains an instruction
      return program, undefined
}
```
I need more clarity on usage of #pragma , currently in algorand documentation,
```
The first line may contain a special version pragma #pragma version X, which directs the assembler to generate TEAL bytecode targeting a certain version. For instance, #pragma version 2 produces bytecode targeting TEAL v2. By default, the assembler targets TEAL v1.
``` 
should we add special function for pragma or ignore it ?

2. we can use `let op = opCodes[words[0]]` this, but i have a question , we need to verify number of fields and data passed from that field. For ex: "addr some_address some_value" this will throw error, should we include a check here ?
there is one more check for address, whether the address is algorand address or not.

Similarly for `int` there are two assert functions.
If we choose to include these asserts in parser function, there will be many assert and switch or if statements in parser function.

That's why in current code they are divided in two functions, pseudo_ops and arithmetic , we can create more function according to opcodes addition but using map may result switch or if conditions.

3.Yes adding line counter is good.

Reply:
Let's add a new opcode for Pragma, which will just store the TEAL version in the interpreter.
BTW: I'm not sure, but I think there can be comments before pragma

Reply:
So, probably we can move argument parsing into the constructor. It could be easier.
Example.. Now we have:
```
export class Store extends Op {
  readonly index: number;
  readonly interpreter: Interpreter;
  constructor (index: number, interpreter: Interpreter) {
```

We can change the constructor to:
```
export class Store extends Op {
  constructor (interpreter: Interpreter, args: string[]) {
    super();
    assert_len(args, 1);
    this.index = parseInt(args[0]);
    this.interpreter = interpreter;
  }
```

BTW, in `let op ... ` there should be `new: `let op = new opCodes[words[0]]`

Reply:
The above is just an idea, but it requires changes in all op codes.
You can try to find a better way if you like.

Reply:
Okay, updating